devito/core/gpu_openmp.py

```diff
@@ -325,41 +326,53 @@ def initialize(iet, **kwargs):
 
     @singledispatch
     def _initialize(iet):
-        return iet
+        return iet, {}
 
     @_initialize.register(EntryFunction)
     def _(iet):
-        comm = None
+        # TODO: we need to pick the rank from `comm_shm`, not `comm`,
+        # so that we have nranks == ngpus (as long as the user has launched
+        # the right number of MPI processes per node given the available
+        # number of GPUs per node)
 
+        objcomm = None
         for i in iet.parameters:
             if isinstance(i, MPICommObject):
-                comm = i
+                objcomm = i
                 break
 
-        if comm is not None:
+        deviceid = DeviceID()
+        if objcomm is not None:
             rank = Symbol(name='rank')
             rank_decl = LocalExpression(DummyEq(rank, 0))
-            rank_init = Call('MPI_Comm_rank', [comm, Byref(rank)])
+            rank_init = Call('MPI_Comm_rank', [objcomm, Byref(rank)])
 
             ngpus = Symbol(name='ngpus')
             call = Function('omp_get_num_devices')()
             ngpus_init = LocalExpression(DummyEq(ngpus, call))
 
-            set_device_num = Call('omp_set_default_device', [rank % ngpus])
-
-            body = [rank_decl, rank_init, ngpus_init, set_device_num]
-
-            init = List(header=c.Comment('Begin of OpenMP+MPI setup'),
-                        body=body,
-                        footer=(c.Comment('End of OpenMP+MPI setup'), c.Line()))
+            osdd_then = Call('omp_set_default_device', [deviceid])
+            osdd_else = Call('omp_set_default_device', [rank % ngpus])
 
-            iet = iet._rebuild(body=(init,) + iet.body)
+            body = [Conditional(
+                CondNe(deviceid, -1),
+                osdd_then,
+                List(body=[rank_decl, rank_init, ngpus_init, osdd_else]),
+            )]
+        else:
+            body = [Conditional(
+                CondNe(deviceid, -1),
+                Call('omp_set_default_device', [deviceid])
+            )]
 
-        return iet
+        init = List(header=c.Comment('Begin of OpenMP+MPI setup'),
+                    body=body,
+                    footer=(c.Comment('End of OpenMP+MPI setup'), c.Line()))
+        iet = iet._rebuild(body=(init,) + iet.body)
```

Comment:
Can't this be merged in a `setup(then, del, ran, ngpu, else)` with the acc one above, unless missread it looks like same call with just different then/else/decl.

Reply:
Actually then/else seem to be the only two different lines.

Reply:
It is indeed feasible. Do you mind if I do it here? https://github.com/devitocodes/devito/blob/refact-op-hierarchy/devito/passes/iet/specializations/openmp.py#L213
this is three branches after this one, and I'm in the middle of a big refactoring as you can see (I've moved and slightly changed the `initialize` methods) so I suggest I do it there. If it's OK with  you I can do it today and push it there directly, this would save some rebasing

Reply:
That's fine there yes, would avoid tracking duplicates omp/acc so would be nice.